x-pack/plugins/stack_connectors/server/connector_types/gemini/gemini.ts

```diff
@@ -0,0 +1,191 @@
+  private static token: string | null;
+  private static tokenExpiryTimeout: NodeJS.Timeout;
+  private url;
+  private model;
+  private gcpRegion;
+  private gcpProjectID;
+
+  constructor(params: ServiceParams<Config, Secrets>) {
+    super(params);
+
+    this.url = this.config.apiUrl;
+    this.model = this.config.defaultModel;
+    this.gcpRegion = this.config.gcpRegion;
+    this.gcpProjectID = this.config.gcpProjectID;
+
+    this.registerSubActions();
+  }
+
+  private registerSubActions() {
+    this.registerSubAction({
+      name: SUB_ACTION.RUN,
+      method: 'runApi',
+      schema: RunActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.DASHBOARD,
+      method: 'getDashboard',
+      schema: DashboardActionParamsSchema,
+    });
+
+    this.registerSubAction({
+      name: SUB_ACTION.TEST,
+      method: 'runApi',
+      schema: RunActionParamsSchema,
+    });
+  }
+
+  protected getResponseErrorMessage(error: AxiosError<{ message?: string }>): string {
+    if (!error.response?.status) {
+      return `Unexpected API Error: ${error.code ?? ''} - ${error.message ?? 'Unknown error'}`;
+    }
+    if (
+      error.response.status === 400 &&
+      error.response?.data?.message === 'The requested operation is not recognized by the service.'
+    ) {
+      return `API Error: ${error.response.data.message}`;
+    }
+    if (error.response.status === 401) {
+      return `Unauthorized API Error${
+        error.response?.data?.message ? `: ${error.response.data.message}` : ''
+      }`;
+    }
+    return `API Error: ${error.response?.statusText}${
+      error.response?.data?.message ? ` - ${error.response.data.message}` : ''
+    }`;
+  }
+
+  /**
+   *  retrieves a dashboard from the Kibana server and checks if the
+   *  user has the necessary privileges to access it.
+   * @param dashboardId The ID of the dashboard to retrieve.
+   */
+  public async getDashboard({
+    dashboardId,
+  }: DashboardActionParams): Promise<DashboardActionResponse> {
+    const privilege = (await this.esClient.transport.request({
+      path: '/_security/user/_has_privileges',
+      method: 'POST',
+      body: {
+        index: [
+          {
+            names: ['.kibana-event-log-*'],
+            allow_restricted_indices: true,
+            privileges: ['read'],
+          },
+        ],
+      },
+    })) as { has_all_requested: boolean };
+
+    if (!privilege?.has_all_requested) {
+      return { available: false };
+    }
+
+    const response = await initDashboard({
+      logger: this.logger,
+      savedObjectsClient: this.savedObjectsClient,
+      dashboardId,
+      genAIProvider: 'Gemini',
+    });
+
+    return { available: response.success };
+  }
+
+  /** Retrieve access token based on the GCP service account credential json file */
+  private async getAccessToken(): Promise<string | null> {
+    const credentials = JSON.parse(this.secrets.credentialsJson);
```

Comment:
we should wrap this in try/catch and warn the user when their `credentialsJson` is not valid JSON

Reply:
done